### PR TITLE
Moving hypothesis and other installations to Docker

### LINKED ISSUE #31136: Move hypothesis installation to docker
Most of our test CI broke today because hypothesis did an update that changed their internal module structure and we were importing things from an internal module. 

In the long-term we should install hypothesis while building our docker images instead of asking our CI to install it every time the CI runs. i.e., remove this line: https://github.com/pytorch/pytorch/blob/7a8261e962f26f84f6211bca4270713a3102146c/.jenkins/pytorch/test.sh#L52 .

I'm going to mark this issue as high-pri because it is related to flaky tests, please correct me if the guidance is incorrect.

cc @ezyang @gchanan @zou3519

### PR DESCRIPTION
**Pull request by janeyx99**
Fixes #31136 

This PR:
1. moves several installations to Docker from `test.sh` for both PyTorch and Caffe2
2. removes version fixing for numba and llvmlite as the issue linked has been resolved

### CODE REVIEW

File: .jenkins/pytorch/common.sh [hunk 1]
```diff
@@ -19,6 +19,7 @@ if [[ "${BUILD_ENVIRONMENT}" == *rocm* ]] && [[ "${BUILD_ENVIRONMENT}" =~ py((2|
   shopt -s expand_aliases
   export PYTORCH_TEST_WITH_ROCM=1
   alias python="$PYTHON"
+  python -m pip install hypothesis # hypothesis was not found in jenkins pytorch-linux-bionic-rocm3.8-py3.6-test1
```
Comment (janeyx99) [inline .jenkins/pytorch/common.sh:22]: Before adding this, the pr/pytorch-linux-bionic-rocm3.8-py3.6 job was failing as the tests could not find the hypothesis package (even though it was installed in Docker). Adding this line fixes it, though there may be a more elegant solution.

Comment (jeffdaily) [reply .jenkins/pytorch/common.sh]: Though the ROCm CI image will be rebuilt successfully by this PR, our ROCm CI is still Jenkins-based and the docker image hash is maintained in the ossci-job-dsl repo.  See https://github.com/pytorch/ossci-job-dsl/pull/78 for an example diff where we update the CI image and hash.

File: .circleci/docker/common/install_conda.sh [hunk 2]
```diff
@@ -98,11 +98,27 @@ if [ -n "$ANACONDA_PYTHON_VERSION" ]; then
...
+    unittest-xml-reporting \
+    coverage \
+    hypothesis==4.53.2 \
+    mypy==0.770 \
+    tb-nightly
+
+  # Update scikit-learn to a python-3.8 compatible version
+  if [[ $(python -c "import sys; print(int(sys.version_info >= (3, 8)))") == "1" ]]; then
+    as_jenkins pip install --progress-bar off -U scikit-learn
+  fi
```
Comment (walterddr) [inline .circleci/docker/common/install_conda.sh:121]: maybe just do:
```suggestion
  if [[ $(python -c "import sys; print(int(sys.version_info >= (3, 8)))") == "1" ]]; then
    as_jenkins pip install --progress-bar off scikit-learn
  else
    as_jenkins pip install --progress-bar off scikit-learn==0.20.3
  fi
```

File: .jenkins/pytorch/common.sh [hunk 3]
```diff
@@ -19,6 +19,8 @@ if [[ "${BUILD_ENVIRONMENT}" == *rocm* ]] && [[ "${BUILD_ENVIRONMENT}" =~ py((2|
   shopt -s expand_aliases
   export PYTORCH_TEST_WITH_ROCM=1
   alias python="$PYTHON"
+  # hypothesis and google was not found in jenkins pytorch-linux-bionic-rocm3.8-py3.6-test1/2
+  python -m pip install hypothesis==4.53.2 protobuf
```
Comment (walterddr) [inline .jenkins/pytorch/common.sh:23]: plz remove and follow @jeffdaily 's suggesiton

File: .jenkins/pytorch/common.sh (hunk 1 above)
Comment (janeyx99) [reply .jenkins/pytorch/common.sh]: Thank you--I've opened a PR [here](https://github.com/pytorch/ossci-job-dsl/pull/80)

**Comment by janeyx99** [comment]
yep rebasing rn to also fix those merge conflicts

**Comment by jeffdaily** [comment]
I'm guessing/assuming the rebase will also change the image tag again.

**Comment by janeyx99** [comment]
hm it might if people have been messing with Dockerfiles

**Comment by walterddr** [comment]
> @pytorchbot retest this please

This has never worked for me before - wondering what was the issue and how did you get it working? (and whether this actually worked as what you intended?)

**Comment by janeyx99** [comment]
> > @pytorchbot retest this please
> 
> This has never worked for me before - wondering what was the issue and how did you get it working? (and whether this actually worked as what you intended?)

I think it worked fo the two tests prefixed by pr/

**Comment by jeffdaily** [comment]
@janeyx99 as far as I'm concerned, you can consider ROCm CI as passing.  An unrelated PR landed that broke ROCm CI, and has since been reverted, but your retest was in the middle of the brokenness.  If you want complete sanity, then you'll have to pytorchbot retest again.  However, I've looked at the CI logs and they're using the correct docker image tag.

**Review by walterddr — APPROVED** [review]

**Pull request closed without merging**